### PR TITLE
Immediate allocation

### PR DESCRIPTION
**Pull request by byako**
Implements #2

### CODE REVIEW

File: api/example.com/resource/gpu/nas/v1alpha1/client/client.go [hunk 1]
```diff
@@ -99,3 +99,17 @@ func (c *Client) Get() error {
 	*c.nas = *crd
 	return nil
 }
+
+func (c *Client) ListNames() ([]string, error) {
+	nasnames := []string{}
+	nases, err := c.client.NodeAllocationStates(c.nas.Namespace).List(
+		context.TODO(), metav1.ListOptions{})
```
Comment (klueska) [inline api/example.com/resource/gpu/nas/v1alpha1/client/client.go:106]: Either leave this all on one line or put all arguments on a new line. Also, `nases` is a weird name, let's just call it `list` or `naslist`, e.g.:
```
	naslist, err := c.client.NodeAllocationStates(c.nas.Namespace).List(
		context.TODO(),
		metav1.ListOptions{})
```

File: api/example.com/resource/gpu/nas/v1alpha1/client/client.go [hunk 2]
```diff
@@ -99,3 +99,17 @@ func (c *Client) Get() error {
...
+	if err != nil {
+		return nasnames, err
```
Comment (klueska) [inline api/example.com/resource/gpu/nas/v1alpha1/client/client.go:108]:
```suggestion
		return nil, err
```

File: api/example.com/resource/gpu/nas/v1alpha1/client/client.go [hunk 3]
```diff
@@ -99,3 +99,17 @@ func (c *Client) Get() error {
...
+	}
+	for _, nas := range nases.Items {
```
Comment (klueska) [inline api/example.com/resource/gpu/nas/v1alpha1/client/client.go:110]:
```suggestion
	for _, nas := range naslist.Items {
```

File: cmd/dra-example-controller/driver.go [hunk 4]
```diff
@@ -70,7 +70,11 @@ func (d driver) GetClassParameters(ctx context.Context, class *resourcev1.Resour
 	return &dc.Spec, nil
 }
 
-func (d driver) GetClaimParameters(ctx context.Context, claim *resourcev1.ResourceClaim, class *resourcev1.ResourceClass, classParameters interface{}) (interface{}, error) {
+func (d driver) GetClaimParameters(
+	ctx context.Context,
+	claim *resourcev1.ResourceClaim,
+	class *resourcev1.ResourceClass,
+	classParameters interface{}) (interface{}, error) {
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:77]: I'm not a fan of wrapping method signatures like this. In any case, this is not relevant to this PR, so it shouldn't be included here. We can have a separate discussion about whether we want to do this or not and then open a PR to address it across the code base.

File: cmd/dra-example-controller/driver.go [hunk 5]
```diff
@@ -81,22 +85,45 @@ func (d driver) GetClaimParameters(ctx context.Context, claim *resourcev1.Resour
 	case gpucrd.GpuClaimParametersKind:
 		gc, err := d.clientset.GpuV1alpha1().GpuClaimParameters(claim.Namespace).Get(ctx, claim.Spec.ParametersRef.Name, metav1.GetOptions{})
 		if err != nil {
-			return nil, fmt.Errorf("error getting GpuClaimParameters called '%v' in namespace '%v': %v", claim.Spec.ParametersRef.Name, claim.Namespace, err)
+			return nil, fmt.Errorf(
+				"error getting GpuClaimParameters called '%v' in namespace '%v': %v",
+				claim.Spec.ParametersRef.Name, claim.Namespace, err)
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:90]: Likewise here

File: cmd/dra-example-controller/driver.go [hunk 6]
```diff
@@ -81,22 +85,45 @@ func (d driver) GetClaimParameters(ctx context.Context, claim *resourcev1.Resour
...
 		}
 		err = d.gpu.ValidateClaimParameters(&gc.Spec)
 		if err != nil {
-			return nil, fmt.Errorf("error validating GpuClaimParameters called '%v' in namespace '%v': %v", claim.Spec.ParametersRef.Name, claim.Namespace, err)
+			return nil, fmt.Errorf(
+				"error validating GpuClaimParameters called '%v' in namespace '%v': %v",
+				claim.Spec.ParametersRef.Name, claim.Namespace, err)
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:96]: And here

File: cmd/dra-example-controller/driver.go [hunk 7]
```diff
@@ -81,22 +85,45 @@ func (d driver) GetClaimParameters(ctx context.Context, claim *resourcev1.Resour
...
 }
 
-func (d driver) Allocate(ctx context.Context, claim *resourcev1.ResourceClaim, claimParameters interface{}, class *resourcev1.ResourceClass, classParameters interface{}, selectedNode string) (*resourcev1.AllocationResult, error) {
+func (d driver) Allocate(
+	ctx context.Context,
+	claim *resourcev1.ResourceClaim,
+	claimParameters interface{},
+	class *resourcev1.ResourceClass,
+	classParameters interface{},
+	selectedNode string) (*resourcev1.AllocationResult, error) {
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:109]: and here....

File: cmd/dra-example-controller/driver.go [hunk 8]
```diff
@@ -81,22 +85,45 @@ func (d driver) GetClaimParameters(ctx context.Context, claim *resourcev1.Resour
...
+
+	return buildAllocationResult(selectedNode, true), nil
+}
+
+func (d driver) allocateDelayedClaim(
+	claim *resourcev1.ResourceClaim,
+	claimParameters interface{},
+	class *resourcev1.ResourceClass,
+	classParameters interface{},
+	selectedNode string) error {
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:126]: Let's undo all of these, and then I'll do another pass on the PR

File: deployments/helm/dra-example-driver/values.yaml [hunk 9]
```diff
@@ -12,7 +12,7 @@ allowDefaultNamespace: false
 imagePullSecrets: []
 image:
   repository: registry.example.com/dra-example-driver
-  pullPolicy: IfNotPresent
+  pullPolicy: Always
```
Comment (klueska) [inline deployments/helm/dra-example-driver/values.yaml:15]: This doesn't seem relevant to this PR -- or is it somehow needed for immediate allocation to work properly?

File: cmd/dra-example-controller/gpu.go [hunk 10]
```diff
@@ -59,6 +59,35 @@ func (g *gpudriver) Allocate(crd *nascrd.NodeAllocationState, claim *resourcev1.
...
 
+func (g *gpudriver) AllocateImmediately(crd *nascrd.NodeAllocationState, ca *controller.ClaimAllocation) error {
+	claimUID := string(ca.Claim.UID)
+
+	cas := []*controller.ClaimAllocation{ca}
+	allocated := g.allocate(crd, cas)
+
+	claimParams := ca.ClaimParameters.(*gpucrd.GpuClaimParametersSpec)
+	if claimParams.Count != len(allocated[claimUID]) {
+		return fmt.Errorf("Not enough resources")
```
Comment (klueska) [inline cmd/dra-example-controller/gpu.go:70]: returned errors should be lower-case

File: cmd/dra-example-controller/driver.go [hunk 11]
```diff
@@ -133,17 +160,91 @@ func (d driver) Allocate(ctx context.Context, claim *resourcev1.ResourceClaim, c
...
+
+	crdconfig := &nascrd.NodeAllocationStateConfig{
+		Namespace: d.namespace,
+		Name:      nodename,
+	}
+
+	crd := nascrd.NewNodeAllocationState(crdconfig)
+
+	client := nasclient.New(crd, d.clientset.NasV1alpha1())
+
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:224]: nit: no new line

File: api/example.com/resource/gpu/nas/v1alpha1/client/client.go (hunk 1 above)
Comment (byako) [reply api/example.com/resource/gpu/nas/v1alpha1/client/client.go:106]: naslist, ok. Happened to move back to one-line with no newlines for arguments before I saw your comments, got convinced by @bart0sh.

File: cmd/dra-example-controller/driver.go [hunk 12]
```diff
@@ -94,9 +94,17 @@ func (d driver) GetClaimParameters(ctx context.Context, claim *resourcev1.Resour
 
 func (d driver) Allocate(ctx context.Context, claim *resourcev1.ResourceClaim, claimParameters interface{}, class *resourcev1.ResourceClass, classParameters interface{}, selectedNode string) (*resourcev1.AllocationResult, error) {
 	if selectedNode == "" {
-		return nil, fmt.Errorf("TODO: immediate allocations not yet supported")
+		return d.allocateImmediateClaim(claim, claimParameters, class, classParameters)
 	}
 
+	if err := d.allocateDelayedClaim(claim, claimParameters, class, classParameters, selectedNode); err != nil {
+		return nil, err
+	}
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:102]: I try and avoid this style of:
```
if err := blah(); err != nil 
```
I can go into a long explanation of why, but the short answer is that it keeps it consistent with how all other calls that return an error along with other variables need to handle their errors. Since this is how it's been done in this codebase elsewhere, let's keep it that way for consistency.

File: cmd/dra-example-controller/driver.go [hunk 13]
```diff
@@ -94,9 +94,17 @@ func (d driver) GetClaimParameters(ctx context.Context, claim *resourcev1.Resour
 
 func (d driver) Allocate(ctx context.Context, claim *resourcev1.ResourceClaim, claimParameters interface{}, class *resourcev1.ResourceClass, classParameters interface{}, selectedNode string) (*resourcev1.AllocationResult, error) {
 	if selectedNode == "" {
-		return nil, fmt.Errorf("TODO: immediate allocations not yet supported")
+		return d.allocateImmediateClaim(claim, claimParameters, class, classParameters)
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:97]: We should wrap the error return here with a:
```
fmt.Errorf("error performing immediate allocation: %v", err)
```

File: cmd/dra-example-controller/driver.go [hunk 14]
```diff
@@ -94,9 +94,17 @@ func (d driver) GetClaimParameters(ctx context.Context, claim *resourcev1.Resour
...
 	}
 
+	if err := d.allocateDelayedClaim(claim, claimParameters, class, classParameters, selectedNode); err != nil {
+		return nil, err
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:101]: We should wrap the error return here with a:
```
fmt.Errorf("error performing delayed allocation: %v", err)
```

File: deployments/helm/dra-example-driver/values.yaml (hunk 9 above)
Comment (byako) [reply deployments/helm/dra-example-driver/values.yaml:15]: It is needed for iterative development, helps pulling fresh build instead of reusing old one, why wouldn't you have it by default?
Strictly speaking it is not related to the immediate allocation functionality, I'll cut it out into separate PR.

File: cmd/dra-example-controller/driver.go (hunk 11 above)
Comment (byako) [reply cmd/dra-example-controller/driver.go:200]: Preceding or trailing?

File: cmd/dra-example-controller/driver.go [hunk 15]
```diff
@@ -94,9 +94,17 @@ func (d driver) GetClaimParameters(ctx context.Context, claim *resourcev1.Resour
...
+	}
+
+	return buildAllocationResult(selectedNode, true), nil
+}
+
+func (d driver) allocateDelayedClaim(claim *resourcev1.ResourceClaim, claimParameters interface{}, class *resourcev1.ResourceClass, classParameters interface{}, selectedNode string) error {
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:107]: It's inconsistent for this to only return an error, but `allocateImmediateClaim()` to return an AllocationResult and an error. They should probably both return an AllocationResult and an error. This should also make the body of the wrapping call simpler and more logical to follow.

File: cmd/dra-example-controller/driver.go (hunk 11 above)
Comment (byako) [reply cmd/dra-example-controller/driver.go:200]: Trailing, never mind.

File: cmd/dra-example-controller/driver.go [hunk 16]
```diff
@@ -133,17 +141,86 @@ func (d driver) Allocate(ctx context.Context, claim *resourcev1.ResourceClaim, c
...
 	}
 
 	onSuccess()
 
-	return buildAllocationResult(selectedNode, true), nil
+	return nil
+}
+
+func (d driver) allocateImmediateClaim(claim *resourcev1.ResourceClaim, claimParameters interface{}, class *resourcev1.ResourceClass, classParameters interface{}) (*resourcev1.AllocationResult, error) {
+
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:158]: erroneous newline

File: cmd/dra-example-controller/driver.go [hunk 17]
```diff
@@ -133,17 +141,86 @@ func (d driver) Allocate(ctx context.Context, claim *resourcev1.ResourceClaim, c
...
+	crdconfig := &nascrd.NodeAllocationStateConfig{
+		Namespace: d.namespace,
+	}
+	crd := nascrd.NewNodeAllocationState(crdconfig)
+
+	client := nasclient.New(crd, d.clientset.NasV1alpha1())
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:164]: This can be pulled up into `Allocate()` with the client passed directly to the `allocateDelayedClaim()` and `allocateImmediateClaim()` calls.

File: cmd/dra-example-controller/gpu.go (hunk 10 above)
Comment (byako) [reply cmd/dra-example-controller/gpu.go]: Done.

File: deployments/helm/dra-example-driver/values.yaml (hunk 9 above)
Comment (byako) [reply deployments/helm/dra-example-driver/values.yaml]: Done.

File: cmd/dra-example-controller/driver.go (hunk 8 above)
Comment (byako) [reply cmd/dra-example-controller/driver.go]: Done.

File: cmd/dra-example-controller/driver.go (hunk 7 above)
Comment (byako) [reply cmd/dra-example-controller/driver.go]: Done.

File: cmd/dra-example-controller/driver.go (hunk 6 above)
Comment (byako) [reply cmd/dra-example-controller/driver.go]: Done.

File: demo/gpu-test-immediate-pod.yaml [hunk 18]
```diff
@@ -0,0 +1,22 @@
+# Pod for gpu-test-immediate-claim.yaml
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: gpu-test-immediate
+  name: pod1
```
Comment (klueska) [inline demo/gpu-test-immediate-pod.yaml:7]: use pod0 for consistency with other tests

File: cmd/dra-example-controller/driver.go (hunk 5 above)
Comment (byako) [reply cmd/dra-example-controller/driver.go]: Done.

File: demo/gpu-test-immediate-pod.yaml [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+  labels:
+    app: pod
+spec:
+  containers:
+  - name: ctr
```
Comment (klueska) [inline demo/gpu-test-immediate-pod.yaml:12]: use ctr0 for consistency with other tests

File: cmd/dra-example-controller/gpu.go [hunk 20]
```diff
@@ -59,6 +59,35 @@ func (g *gpudriver) Allocate(crd *nascrd.NodeAllocationState, claim *resourcev1.
 	return onSuccess, nil
 }
 
+func (g *gpudriver) AllocateImmediately(crd *nascrd.NodeAllocationState, ca *controller.ClaimAllocation) error {
```
Comment (klueska) [inline cmd/dra-example-controller/gpu.go:62]: If we call this one `AllocateImmediately` can we call the other one `AllocateDelayed` for consistency? Also, is there a reason to not keep the function signatures identical (possibly with the only difference being the removal of `selectedNode`?

File: cmd/dra-example-controller/gpu.go [hunk 21]
```diff
@@ -59,6 +59,35 @@ func (g *gpudriver) Allocate(crd *nascrd.NodeAllocationState, claim *resourcev1.
...
+			UUID: gpu,
+		}
+		devices = append(devices, device)
+	}
+
+	requestedDevices := nascrd.RequestedDevices{
+		Gpu: &nascrd.RequestedGpus{
+			Devices: devices,
+		},
+	}
```
Comment (klueska) [inline cmd/dra-example-controller/gpu.go:85]: Can we factor this out to a function so that it can be reused in unsuitable nodes and here?

 Ideally this function and unsuitable nodes would be almost line-for-line identical (and very short) except for putting the request into `PendingClaimRequests`.

File: cmd/dra-example-controller/driver.go [hunk 22]
```diff
@@ -133,17 +141,85 @@ func (d driver) Allocate(ctx context.Context, claim *resourcev1.ResourceClaim, c
...
+		return nil, fmt.Errorf("could not find suitable devices: %v", err)
+	}
+
+	// sync ResourceClaimAllocations from allocations
+	err = client.Update(&crd.Spec)
+	if err != nil {
+		return nil, fmt.Errorf("updating NodeAllocationState CRD: %v", err)
+	}
+
+	d.gpu.PendingClaimRequests.Remove(string(ca.Claim.UID))
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:220]: Why is this necessary? When / how is this claim ending up in here?

File: cmd/dra-example-controller/driver.go [hunk 23]
```diff
@@ -133,17 +141,85 @@ func (d driver) Allocate(ctx context.Context, claim *resourcev1.ResourceClaim, c
...
+	for _, nodename := range nasnames {
+		if allocationResult, err := d.allocateImmediateClaimOnNode(ca, nodename); err == nil {
+			return allocationResult, nil
+		}
+	}
+
+	return nil, fmt.Errorf("no suitable node found")
+}
+
+func (d driver) allocateImmediateClaimOnNode(ca *controller.ClaimAllocation, nodename string) (*resourcev1.AllocationResult, error) {
```
Comment (klueska) [inline cmd/dra-example-controller/driver.go:188]: Something doesn't feel right with all the redundancies in this function (especially around creating a new client). It feels like some other abstraction / refactoring would be better, but I don't have a suggestion off the top of my head without digging into the code myself and playing with it.

File: demo/gpu-test-immediate-pod.yaml (hunk 19 above)
Comment (byako) [reply demo/gpu-test-immediate-pod.yaml:12]: And pod0, yes, thanks.

File: cmd/dra-example-controller/gpu.go (hunk 20 above)
Comment (byako) [reply cmd/dra-example-controller/gpu.go:62]: Sure, but can we at least wrap claim, claimParams, class and classParams into ClaimAllocation right in the beginning of Allocate and pass it into allocateDelayedClaim and allocateImmedateClaim, so we don't have 160 chars long parameters line in func definition?

Comment (byako) [reply cmd/dra-example-controller/gpu.go:62]: No it has to be broken down again before being passed into gpu module, never mind, I'm just underslept, sorry.

File: api/example.com/resource/gpu/nas/v1alpha1/client/client.go [hunk 24]
```diff
@@ -99,3 +99,17 @@ func (c *Client) Get() error {
 	*c.nas = *crd
 	return nil
 }
+
+func (c *Client) ListNames() ([]string, error) {
+	nasnames := []string{}
```
Comment (bart0sh) [inline api/example.com/resource/gpu/nas/v1alpha1/client/client.go:104]:
```suggestion
```

File: api/example.com/resource/gpu/nas/v1alpha1/client/client.go [hunk 25]
```diff
@@ -99,3 +99,17 @@ func (c *Client) Get() error {
...
+	naslist, err := c.client.NodeAllocationStates(c.nas.Namespace).List(
+		context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
```
Comment (bart0sh) [inline api/example.com/resource/gpu/nas/v1alpha1/client/client.go:109]: nit:
```suggestion
	}
	nasnames := make([]string, 0, len(naslist.Items))
```

File: cmd/dra-example-controller/gpu.go (hunk 21 above)
Comment (byako) [reply cmd/dra-example-controller/gpu.go:85]: Done.

File: cmd/dra-example-controller/gpu.go (hunk 20 above)
Comment (byako) [reply cmd/dra-example-controller/gpu.go:62]: Done.

File: cmd/dra-example-controller/driver.go (hunk 15 above)
Comment (byako) [reply cmd/dra-example-controller/driver.go:107]: That's true. Both of them now return allocation result.

File: cmd/dra-example-controller/driver.go (hunk 14 above)
Comment (byako) [reply cmd/dra-example-controller/driver.go:101]: Done.

File: cmd/dra-example-controller/driver.go (hunk 13 above)
Comment (byako) [reply cmd/dra-example-controller/driver.go:97]: Done.

File: cmd/dra-example-controller/driver.go (hunk 12 above)
Comment (byako) [reply cmd/dra-example-controller/driver.go:102]: Sure, alright.

File: cmd/dra-example-controller/gpu.go [hunk 26]
```diff
@@ -150,3 +160,22 @@ func (g *gpudriver) allocate(crd *nascrd.NodeAllocationState, pod *corev1.Pod, g
 
 	return allocated
 }
+
+func (g *gpudriver) buildRequestedGpuDevices(allocatedDevices []string) nascrd.RequestedDevices {
+
+	var devices []nascrd.RequestedGpu
```
Comment (bart0sh) [inline cmd/dra-example-controller/gpu.go:166]: nit: you can use make with capacity here

File: cmd/dra-example-controller/driver.go [hunk 27]
```diff
@@ -93,10 +93,24 @@ func (d driver) GetClaimParameters(ctx context.Context, claim *resourcev1.Resour
...
+		allocationResult, err := d.allocateImmediateClaim(claim, claimParameters, class, classParameters)
+		if err != nil {
+			return nil, fmt.Errorf("error performing immediate allocation: %v", err)
+		}
+		return allocationResult, nil
 	}
 
+	allocationResult, err := d.allocateDelayedClaim(claim, claimParameters, class, classParameters, selectedNode)
+	if err != nil {
+		return nil, fmt.Errorf("error performing delayed allocation: %v", err)
```
Comment (bart0sh) [inline cmd/dra-example-controller/driver.go:107]: Can we just trust the code of local functions and simply return errors as they are? This also applies to the `allocateImmediateClaim ` call above.

Comment (klueska) [reply cmd/dra-example-controller/driver.go:107]: It's not about "trusting" it. It's about consistency and being able to trace the errors through the function calls when you see them appear in the logs.

Comment (klueska) [reply cmd/dra-example-controller/driver.go:107]: I specifically requested @byako to add these wrappers

File: api/example.com/resource/gpu/nas/v1alpha1/client/client.go [hunk 28]
```diff
@@ -99,3 +99,22 @@ func (c *Client) Get() error {
...
 	return nil
 }
+
+func (c *Client) Switch(newcrd *nascrd.NodeAllocationState) {
+	*c.nas = *newcrd
+}
+
+func (c *Client) GetAll() (map[string]*nascrd.NodeAllocationState, error) {
+	naslist, err := c.client.NodeAllocationStates(c.nas.Namespace).List(
+		context.TODO(), metav1.ListOptions{})
```
Comment (bart0sh) [inline api/example.com/resource/gpu/nas/v1alpha1/client/client.go:109]: Does this line differ somehow? Why do we wrap?

Comment (byako) [reply api/example.com/resource/gpu/nas/v1alpha1/client/client.go:109]: Missed this when unwrapping. Wiill fix with and rebase to renaming change

**Comment by byako** [comment]
Accomodated for renamed CRD fields.

**Comment by asm582** [comment]
are we planning to support `GpuClaimParameters` in immediate mode?

**Comment by byako** [comment]
There's been quite some changes since this PR was submitted, I think it'd be easier to close this one and submit this code rebased and slightly refactored for Kubernetes 1.28 API as a new PR.

**Comment by byako** [comment]
I can pick this up later this week if nobody else will do so by then.

**Comment by klueska** [comment]
I don't think we need to close the PR -- just rebase this one.

**Comment by klueska** [comment]
> For this PR I set the GPU count as 8:
> 
> ```
> apiVersion: gpu.resource.example.com/v1alpha1
> kind: GpuClaimParameters
> metadata:
>   namespace: gpu-test4
>   name: multiple-gpus
> spec:
>   count: 8
> ```
> 
> We see that only 4 claims are allocated:
> 
> ```
>   Allocated Claims:
>   b21d90f4-d444-427e-84ad-196575e2f05f:
>     Gpu:
>       Devices:
>         Uuid:  GPU-657bd2e7-f5c2-a7f2-fbaa-0d1cdc32f81b
>         Uuid:  GPU-ee3e4b55-fcda-44b8-0605-64b7a9967744
>         Uuid:  GPU-9ede7e32-5825-a11b-fa3d-bab6d47e0243
>         Uuid:  GPU-18db0e85-99e9-c746-8531-ffeb86328b39
> ```
> 
> not sure what I am missing here?

Are you sure you didn't accidentally run with an old `GpuClaimParameters` object with count set to `4`? It won't necessarily be updated using ´kubectl apply` if the object already existed with a different value set for `count`.
